keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,660 @@
+Kubernetes to handle validation of string fields for cases where regex is poorly
+suited or insufficient.
+- Immutability
+- Unions
+
+These improvements are largely complementary to expression support and either
+are (or should be) addressed by in separate KEPs.
+
+For use cases cannot be covered by build-in validation support:
+
+- Admission Webhooks: have validating admission webhook for further validation
+- Custom validators: write custom checks in several languages such as Rego 
+
+### Descriptive, self contained CRDs
+
+This KEP will make CRDs more self-contained. Instead of having
+validation rules coded into webhooks that must be
+registered and upgraded independent of a CRD, the rules will be contained within
+the CRD object definition, making them easier to author and introspect by
+cluster administrators and users, and eliminating version skew issues that can
+happen between a CRD and webhook since they can be registered and
+upgraded/rolled-back independently.
+
+### Webhooks: Development Complexity
+
+Introducing a production grade webhook is a substantial development task.
+Beyond authoring the actual core logic that a webhook must perform, the webhook
+must be instrumented for monitoring and alerting and integrated with the
+packaging/releases processes for the environments it will be run it.
+
+The developer must also carefully consider the upgrade and rollback ordering
+between the webhook and CRD.
+
+### Webhooks: Operational Complexity
+
+Admission webhooks are part of the critical serving path of the kube-apiserver.
+Admission webhooks add latency to requests, and large numbers of webhooks
+can cause, or contribute to, request timeouts being exceeded.
+
+Webhooks must either be configured as `FailPolicy.Fail` or `FailPolicy.Ignore`. If
+`FailPolicy.Ignore` is used, there is potential for requests skip the webhook and
+be admitted. If `FailPolicy.Fail` is used, a webhook outage can result in a
+localized or widespread Kubernetes control plane outage depending on which
+objects the webhook is configured to intercept.
+
+### Goals
+
+- Make CRDs more self-contained and declarative
+- Simplify CRD development
+- Simplify CRD operations
+
+### Non-Goals
+
+- Support for validation formats, immutability or unions. These are all valuable improvements
+  but can be addressed orthogonally in separate KEPs.
+- Eliminate the need for webhooks entirely. Webhooks will still be needed for
+  some use cases. For example, if a validation check requires making a network
+  request to some other system, it will still need to be implemented in a webhook.
+- Support all validation done on native Kubernetes types. For example, CRD structural schemas has
+  complex validation rules that we CEL cannot support due to the lask of support for arbitrarily
+  deeply nested terms (CEL cannot support recursive data types).
+- Change how Kubernetes built-in types are validated, defaulted and converted.
```

Comment:
Scoping this KEP to CRDs makes sense, especially for conversion bits

For validation, I really want the ability to evaluate CEL expressions on any type (including built-in types) instead of using webhooks for all of the same operational complexity and reliability reasons stated above.

As a thought experiment, if custom admission could use CEL expressions against any resource, would it still make sense to bake CEL expressions into the CRD validation schema as this proposes, or would the general admission mechanism be sufficient?

Reply:
This is really helpful. I've incorporated it into the KEP. I've tried to retain some attribution and links back to these this comment.